zpy/client_util.py

```diff
@@ -123,3 +136,225 @@ def clear_last_print():
 def is_done(state: str):
     """Returns True if state is a done state, false otherwise."""
     return state in ["READY", "CANCELLED", "PACKAGING_FAILED", "GENERATING_FAILED"]
+
+
+def unique_list(list: list) -> list:
+    """Removes non unique items from a list. Works for objects, unlike set()"""
+    return [i for n, i in enumerate(list) if i not in list[n + 1 :]]
+
+
+def dict_to_list(d: dict) -> list:
+    """Converts dict to list"""
+    return list(dict(d).values())
+
+
+def remove_n_extensions(path: Union[str, Path], n: int) -> Path:
+    """Takes a path and removes n extensions from the end. Example: "image.rgb.png" becomes "image" for n = 2"""
+    p = Path(path)
+    extensions = "".join(p.suffixes[-n:])  # remove n extensions
+    return str(p).removesuffix(extensions)
+
+
+def extract_zip(path_to_zip: Path) -> Path:
+    """
+    Extracts a .zip to a new adjacent folder by the same name.
+    Args:
+        path_to_zip: Path to .zip
+    Returns:
+        Path: Extracted folder path
+    """
+    unzipped_path = Path(remove_n_extensions(path_to_zip, n=1))
+    with zipfile.ZipFile(path_to_zip, "r") as zip_ref:
+        zip_ref.extractall(unzipped_path)
+    return unzipped_path
+
+
+def group_metadata_by_datapoint(dataset_path: Path) -> list[dict]:
+    """
+    Updates metadata with new ids and accurate image paths.
+    Returns a list of dicts, each item containing metadata relevant to a single datapoint.
+    Args:
+        dataset_path (Path): Path to unzipped dataset.
+    Returns:
+        list[
+            dict[
+                metadata: list[dict],
+                categories: list[dict],
+                images: list[dict],
+                annotations: list[dict],
+
+            ]
+        ]: Returns a list of dicts, each item containing metadata relevant to a single datapoint.
+    """
+
+    metadata_dicts = []
+    # batch level - sum category counts
+    category_count_sums = {}
+    for batch in listdir(dataset_path):
+        batch_uri = join(dataset_path, batch)
+        annotation_file_uri = join(batch_uri, "_annotations.zumo.json")
+        metadata = json.load(open(annotation_file_uri))
+        batch_categories = dict_to_list(metadata["categories"])
+        for c in batch_categories:
+            category_count_sums[c["id"]] = (
+                category_count_sums.get(c["id"], 0) + c["count"]
+            )
+
+    # batch level - group images by satapoint
+    for batch in listdir(dataset_path):
```

Comment:
I can't because the category counts have to be summed beforehand, otherwise the category count property will be 10, 20, 30, 40, 50 depending on the batch

Reply:
Edit: I guess I can do this now since the function is accumulating something rather than calling the callback directly from the datapoint iterations.